spec/structured-data.html

```diff
@@ -0,0 +1,131 @@
+            1. If IsCallable(_toJSON_) is *true*, then
+              1. Set _value_ to ? Call(_toJSON_, _value_, &laquo; _key_ &raquo;).
+          1. If _state_.[[ReplacerFunction]] is not *undefined*, then
+            1. Set _value_ to ? Call(_state_.[[ReplacerFunction]], _holder_, &laquo; _key_, _value_ &raquo;).
+          1. If Type(_value_) is Object, then
+            1. If _value_ has a [[NumberData]] internal slot, then
+              1. Set _value_ to ? ToNumber(_value_).
+            1. Else if _value_ has a [[StringData]] internal slot, then
+              1. Set _value_ to ? ToString(_value_).
+            1. Else if _value_ has a [[BooleanData]] internal slot, then
+              1. Set _value_ to _value_.[[BooleanData]].
+            1. Else if _value_ has a [[BigIntData]] internal slot, then
+              1. Set _value_ to _value_.[[BigIntData]].
+          1. If _value_ is *null*, return *"null"*.
+          1. If _value_ is *true*, return *"true"*.
+          1. If _value_ is *false*, return *"false"*.
+          1. If Type(_value_) is String, return QuoteJSONString(_value_).
+          1. If Type(_value_) is Number, then
+            1. If _value_ is finite, return ! ToString(_value_).
+            1. Return *"null"*.
+          1. If Type(_value_) is BigInt, throw a *TypeError* exception.
+          1. If Type(_value_) is Object and IsCallable(_value_) is *false*, then
+            1. Let _isArray_ be ? IsArray(_value_).
+            1. If _isArray_ is *true*, return ? SerializeJSONArray(_state_, _value_).
+            1. Return ? SerializeJSONObject(_state_, _value_).
+          1. <ins>If Type(_value_) is Record, return ? SerializeJSONObject(_state_, _value_).</ins>
+          1. <ins>If Type(_value_) is Tuple, return ? SerializeJSONArray(_state_, _value_).</ins>
+          1. Return *undefined*.
+        </emu-alg>
+      </emu-clause>
+
+      <emu-clause id="sec-serializejsonobject" aoid="SerializeJSONObject">
+        <h1>Runtime Semantics: SerializeJSONObject ( _state_, _value_ )</h1>
+        <p>The abstract operation SerializeJSONObject takes arguments _state_ and _value_. It serializes an object <ins>or a Record to a JSON object</ins>. It performs the following steps when called:</p>
+        <emu-alg>
+          1. <ins>Assert: Either Type(_value_) is Record or Type(_value_) is Object.</ins>
+          1. If _state_.[[Stack]] contains _value_, throw a *TypeError* exception because the structure is cyclical.
+          1. Append _value_ to _state_.[[Stack]].
+          1. Let _stepback_ be _state_.[[Indent]].
+          1. Set _state_.[[Indent]] to the string-concatenation of _state_.[[Indent]] and _state_.[[Gap]].
+          1. If _state_.[[PropertyList]] is not *undefined*, then
+            1. Let _K_ be _state_.[[PropertyList]].
+          1. <ins>Else if Type(_value_) is Record, then</ins>
+            1. <ins>Let _K_ be _value_.[[OwnPropertyKeys]]().</ins>
```

Comment:
> I think we should instead fix EnumerableOwnPropertyNames to work on Records. This will be important for Object.keys to work on Records. (I didn't realize this was broken before doing this review.) If we were going to not do that, note that it's necessary to create a Record wrapper object and call [[OwnPropertyKeys]] on that; it doesn't make sense to call this method on the primitive value.

`Object.keys` and the other methods should already work with records, because they all call `ToObject`. It's only an internal problem in the spec, and not in the public interface.

> :+1:

Reply:
Well, this callsite still needs to be fixed.